tests/document/test_instance.py

```diff
@@ -3911,6 +3911,13 @@ def test_dbfields_are_loaded_to_the_right_modelfield_for_dyn_doc_2(self):
             reloaded.z2,
         ) == (doc.x1, doc.x2, doc.y1, doc.y2, doc.z1, doc.z2)
 
+    def test_from_son_with_compiled_regex(self):
+        class User(DynamicDocument):
+            re_str = StringField(regex=r"(^ABC\d\d\d\d$)")
+
+        data = {"re_str": "ABC0000"}
+        User._from_son(son=data, _auto_dereference=False)
+
 
```

Comment:
I'm not used to overwriting deepcopy but from what I read, there are known risks of recursion problem when overwriting deepcopy, so it would be great to have a test that covers that and make sure it doesn't loop infinitely. Could you also add another test with a User having a a reference to itself, something like

```
import copy

class User(Document):
    name = StringField(regex=r"(.*)")
    other_user = ReferenceField('self')

user1 = User(name='John').save()
user2 = User(name='Bob', other_user=user1).save()

user1.other_user = user1
user1.save()

for us in User.objects():
    copied_user = copy.deepcopy(us)
    assert copied_user is not us
    assert copied_user.regex is not us.regex
```

Since that new test would relate to StringField more than instance, that one would make sense to be in test_string_field.

Reply:
Sure, I'll try to add more test.